jolt-core/benches/grand_product.rs

```diff
@@ -0,0 +1,239 @@
+
+const SRS_SIZE: usize = 1 << 10;
+
+#[derive(Clone, Copy)]
+struct BenchConfig {
+    pub name: &'static str,
+    pub num_layers: usize,
+    pub layer_size: usize,
+    pub threshold: u32,
+}
+
+// Sets up the benchmark by generating leaves and computing known products
+// and allows configuring the percentage of ones in the leaves
+fn setup_bench<PCS, F>(
+    num_batches: usize,
+    layer_size: usize,
+    threshold: u32,
+) -> (
+    // Leaves
+    Vec<Vec<F>>,
+    PCS::Setup,
+    // Products of leaves
+    Vec<F>,
+)
+where
+    PCS: CommitmentScheme<Field = F>,
+    F: JoltField,
+{
+    assert!(
+        threshold <= 100,
+        "Threshold must be between 0 and 100, but got {}",
+        threshold
+    );
+
+    let mut rng = ChaCha20Rng::seed_from_u64(111111u64);
+
+    let threshold = ((threshold as u64 * u32::MAX as u64) / 100) as u32;
+
+    // Generate leaves with percentage of ones
+    let leaves: Vec<Vec<F>> = (0..num_batches)
+        .map(|_| {
+            (0..layer_size)
+                .map(|_| {
+                    if rng.next_u32() < threshold {
+                        F::one()
+                    } else {
+                        F::random(&mut rng)
+                    }
+                })
+                .collect()
+        })
+        .collect();
+
+    // Compute known products (one per layer)
+    let known_products: Vec<F> = leaves.iter().map(|layer| layer.iter().product()).collect();
+
+    let setup = PCS::setup(&[CommitShape::new(SRS_SIZE, BatchType::Big)]);
+
+    (leaves, setup, known_products)
+}
+
+fn benchmark_prove<PCS, F, G>(
+    c: &mut Criterion,
+    config: BenchConfig,
+    grand_products_config: G::Config,
+) where
+    PCS: CommitmentScheme<Field = F>,
+    F: JoltField,
+    G: BatchedGrandProduct<F, PCS, Leaves = Vec<Vec<F>>>,
+{
+    let (leaves, setup, _) =
+        setup_bench::<PCS, F>(config.num_layers, config.layer_size, config.threshold);
+
+    let mut grand_product = G::construct_with_config(leaves, grand_products_config);
+
+    c.bench_function(
+        &format!(
+            "Grand Product Prove: {} - {}% Ones",
+            config.name, config.threshold
+        ),
+        |b| {
+            b.iter(|| {
+                // Prove the grand product
+                let mut transcript = ProofTranscript::new(b"test_transcript");
+                let mut prover_accumulator: ProverOpeningAccumulator<F> =
+                    ProverOpeningAccumulator::new();
+                let _proof: BatchedGrandProductProof<PCS> = grand_product
+                    .prove_grand_product(
+                        Some(&mut prover_accumulator),
+                        &mut transcript,
+                        Some(&setup),
+                    )
+                    .0;
+
+                if prover_accumulator.len() > 0 {
+                    let _ =
+                        Some(prover_accumulator.reduce_and_prove::<PCS>(&setup, &mut transcript));
+                }
```

Comment:
let's pull this out of the bench, as the cost of the opening proof is (mostly) amortized across all openings in Jolt